{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Verify Data Fields\n",
    "2023-04-05 ZD  \n",
    "\n",
    "This notebook will compare high-level data elements between versions of data received from CHoP for use in MTP."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import modules and define relative paths\n",
    "Data structure will be based on the `verify_data_displayed_in_mtp.ipynb` DV3 Notebook. Edit core older/newer versions and all other file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Core Older Versions for Comparison\n",
    "OLD_OPENPEDCAN_SOMATIC_ALTERATIONS_VERSION = 'v12pre'\n",
    "OLD_OPENPEDCAN_GENE_EXPRESSION_VERSION = 'v12pre'\n",
    "OLD_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION = 'v12pre'\n",
    "\n",
    "# Core Newer Versions for Comparison\n",
    "NEW_OPENPEDCAN_SOMATIC_ALTERATIONS_VERSION = 'v12pre2'\n",
    "NEW_OPENPEDCAN_GENE_EXPRESSION_VERSION = 'v12pre2'\n",
    "NEW_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION = 'v12pre2'\n",
    "\n",
    "# Do not change below\n",
    "# ------\n",
    "\n",
    "# Static CHoP: Epigenetic Modification (Methylation)\n",
    "# Does not change between versions\n",
    "METHYL_FILENAME = 'isoform-methyl-beta-values-summary'\n",
    "METHYLGENE_FILENAME = 'gene-methyl-beta-values-summary'\n",
    "\n",
    "# ---\n",
    "\n",
    "# Older Data from CHoP\n",
    "OLD_CHOP_SA_PATH = 'data/raw/chopOpenPedCan/somaticAlterations/' + OLD_OPENPEDCAN_SOMATIC_ALTERATIONS_VERSION + '/'\n",
    "OLD_CHOP_GX_PATH = 'data/raw/chopOpenPedCan/geneExpression/' + OLD_OPENPEDCAN_GENE_EXPRESSION_VERSION + '/'\n",
    "OLD_CHOP_EM_PATH = 'data/raw/chopOpenPedCan/epigeneticModification/' + OLD_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION + '/'\n",
    "\n",
    "# Older CHoP: Somatic Alterations\n",
    "OLD_CNV_PATH = OLD_CHOP_SA_PATH + 'gene-level-cnv-consensus-annotated-mut-freq.jsonl.gz'\n",
    "OLD_SNVGENE_PATH = OLD_CHOP_SA_PATH + 'gene-level-snv-consensus-annotated-mut-freq.jsonl.gz'\n",
    "OLD_SNV_PATH = OLD_CHOP_SA_PATH + 'variant-level-snv-consensus-annotated-mut-freq.jsonl.gz'\n",
    "OLD_FUSIONGENE_PATH = OLD_CHOP_SA_PATH + 'putative-oncogene-fused-gene-freq.jsonl.gz'\n",
    "OLD_FUSION_PATH = OLD_CHOP_SA_PATH + 'putative-oncogene-fusion-freq.jsonl.gz'\n",
    "\n",
    "# Older CHoP: Gene Expression\n",
    "OLD_TPMGENE_PATH = OLD_CHOP_GX_PATH + 'long_n_tpm_mean_sd_quantile_gene_wise_zscore.jsonl.gz'\n",
    "OLD_TPMGROUP_PATH = OLD_CHOP_GX_PATH + 'long_n_tpm_mean_sd_quantile_group_wise_zscore.jsonl.gz'\n",
    "\n",
    "# Older CHoP: Raw Methylation (large files)\n",
    "OLD_METHYL_PATH = OLD_CHOP_EM_PATH + METHYL_FILENAME + '.jsonl.gz'\n",
    "OLD_METHYLGENE_PATH = OLD_CHOP_EM_PATH + METHYLGENE_FILENAME + '.jsonl.gz'\n",
    "\n",
    "# Older CHoP: Grouped Methylation (aggregated small files)\n",
    "OLD_GROUPED_CHOP_EM_PATH = 'data/processed/chopOpenPedCan/epigeneticModification_grouped/' + OLD_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION + '/'\n",
    "OLD_GROUPED_METHYL_PATH = OLD_GROUPED_CHOP_EM_PATH + METHYL_FILENAME + '/'\n",
    "OLD_GROUPED_METHYLGENE_PATH = OLD_GROUPED_CHOP_EM_PATH + METHYLGENE_FILENAME + '/'\n",
    "\n",
    "# ------\n",
    "\n",
    "# Newer Data from CHoP \n",
    "NEW_CHOP_SA_PATH = 'data/raw/chopOpenPedCan/somaticAlterations/' + NEW_OPENPEDCAN_SOMATIC_ALTERATIONS_VERSION + '/'\n",
    "NEW_CHOP_GX_PATH = 'data/raw/chopOpenPedCan/geneExpression/' + NEW_OPENPEDCAN_GENE_EXPRESSION_VERSION + '/'\n",
    "NEW_CHOP_EM_PATH = 'data/raw/chopOpenPedCan/epigeneticModification/' + NEW_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION + '/'\n",
    "\n",
    "# Newer CHoP: Somatic Alterations\n",
    "NEW_CNV_PATH = NEW_CHOP_SA_PATH + 'gene-level-cnv-consensus-annotated-mut-freq.jsonl.gz'\n",
    "NEW_SNVGENE_PATH = NEW_CHOP_SA_PATH + 'gene-level-snv-consensus-annotated-mut-freq.jsonl.gz'\n",
    "NEW_SNV_PATH = NEW_CHOP_SA_PATH + 'variant-level-snv-consensus-annotated-mut-freq.jsonl.gz'\n",
    "NEW_FUSIONGENE_PATH = NEW_CHOP_SA_PATH + 'putative-oncogene-fused-gene-freq.jsonl.gz'\n",
    "NEW_FUSION_PATH = NEW_CHOP_SA_PATH + 'putative-oncogene-fusion-freq.jsonl.gz'\n",
    "\n",
    "# Newer CHoP: Gene Expression\n",
    "NEW_TPMGENE_PATH = NEW_CHOP_GX_PATH + 'long_n_tpm_mean_sd_quantile_gene_wise_zscore.jsonl.gz'\n",
    "NEW_TPMGROUP_PATH = NEW_CHOP_GX_PATH + 'long_n_tpm_mean_sd_quantile_group_wise_zscore.jsonl.gz'\n",
    "\n",
    "# Newer CHoP: Raw Methylation (large files)\n",
    "NEW_METHYL_PATH = NEW_CHOP_EM_PATH + METHYL_FILENAME + '.jsonl.gz'\n",
    "NEW_METHYLGENE_PATH = NEW_CHOP_EM_PATH + METHYLGENE_FILENAME + '.jsonl.gz'\n",
    "\n",
    "# Newer CHoP: Grouped Methylation (aggregated small files)\n",
    "NEW_GROUPED_CHOP_EM_PATH = 'data/processed/chopOpenPedCan/epigeneticModification_grouped/' + NEW_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION + '/'\n",
    "NEW_GROUPED_METHYL_PATH = NEW_GROUPED_CHOP_EM_PATH + METHYL_FILENAME + '/'\n",
    "NEW_GROUPED_METHYLGENE_PATH = NEW_GROUPED_CHOP_EM_PATH + METHYLGENE_FILENAME + '/'\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare data versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_jsonl_as_chunks(path:str, limit=None, chunksize=2.5e5):\n",
    "    \"\"\"\n",
    "    Load large jsonl file in chunks and then combine to save memory.\n",
    "\n",
    "    :param path: Relative filepath to jsonl file\n",
    "    :param limit: max number of chunks to include in concat. Use to\n",
    "        load a subset of the data\n",
    "    :param chunksize: int size of the lines to include\n",
    "\n",
    "    \"\"\"\n",
    "    # Load jsonl in chunks\n",
    "    df_chunks = pd.read_json(path, orient='records', lines=True, chunksize=chunksize)\n",
    "\n",
    "    # Create empty dataframe\n",
    "    df = pd.DataFrame()\n",
    "\n",
    "    # Fill empty dataframe with chunks up to limit\n",
    "    for label, chunk in enumerate(df_chunks):\n",
    "        if label == limit:\n",
    "            break\n",
    "        df = pd.concat([df,chunk], ignore_index=True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tuples of (old, new) paths for iteration\n",
    "pathTuples = [\n",
    "            (OLD_CNV_PATH,          NEW_CNV_PATH),\n",
    "            (OLD_SNVGENE_PATH,      NEW_SNVGENE_PATH),\n",
    "            (OLD_SNV_PATH,          NEW_SNV_PATH),\n",
    "            (OLD_FUSIONGENE_PATH,   NEW_FUSIONGENE_PATH),\n",
    "            (OLD_FUSION_PATH,       NEW_FUSION_PATH),\n",
    "            (OLD_TPMGENE_PATH,      NEW_TPMGENE_PATH),\n",
    "            (OLD_TPMGROUP_PATH,     NEW_TPMGROUP_PATH),\n",
    "            (OLD_METHYL_PATH,       NEW_METHYL_PATH),\n",
    "            (OLD_METHYLGENE_PATH,   NEW_METHYLGENE_PATH)\n",
    "            ]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check fields present across file versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output version descriptions\n",
    "introText = (\n",
    "    f'Comparing data fields between versions:\\n---\\n'\n",
    "    f'    Somatic Alterations:        {OLD_OPENPEDCAN_SOMATIC_ALTERATIONS_VERSION:8}'\n",
    "                            f'-->     {NEW_OPENPEDCAN_SOMATIC_ALTERATIONS_VERSION:8}\\n'\n",
    "    f'    Gene Expression:            {OLD_OPENPEDCAN_GENE_EXPRESSION_VERSION:8}'\n",
    "                            f'-->     {NEW_OPENPEDCAN_GENE_EXPRESSION_VERSION:8}\\n'\n",
    "    f'    Epigenetic Modification:    {OLD_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION:8}'\n",
    "                            f'-->     {NEW_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION:8}\\n')\n",
    "print(introText, '---\\n')\n",
    "\n",
    "# Iterate though data version groups\n",
    "for oldPath, newPath in pathTuples:\n",
    "\n",
    "    # Get filename(s) for output labeling\n",
    "    filename = oldPath.rsplit('/',1)[-1]\n",
    "    if filename != newPath.rsplit('/',1)[-1]:\n",
    "        filename = oldPath.rsplit('/',1)[-1]+' renamed '+newPath.rsplit('/',1)[-1]\n",
    "\n",
    "    print(f'{filename} ...')\n",
    "\n",
    "    # If new or old file is missing, skip comparison\n",
    "    if not os.path.exists(oldPath) or not os.path.exists(newPath):\n",
    "        print(f'    Comparison not available - File(s) not found.')\n",
    "        continue\n",
    "    \n",
    "    # Load first 10,000 lines of each dataset to gather columns\n",
    "    # (Loading more than 1 to avoid dropping columns with NaN values)\n",
    "    old = load_jsonl_as_chunks(oldPath, limit=1, chunksize=1e4)\n",
    "    new = load_jsonl_as_chunks(newPath, limit=1, chunksize=1e4)\n",
    "\n",
    "    # Find different columns between versions\n",
    "    old_unique = list(set(old.columns.tolist()) - set(new.columns.tolist()))\n",
    "    new_unique = list(set(new.columns.tolist()) - set(old.columns.tolist()))\n",
    "    \n",
    "    # Print output with filename label if any have mismatches\n",
    "    if len(new_unique) + len(old_unique) == 0:\n",
    "        print(f'    All good. No field changes.')\n",
    "    if len(new_unique) > 0:\n",
    "        print('    New fields added:', *new_unique, sep='\\n        ')\n",
    "    if len(old_unique) > 0:\n",
    "        print('    Old fields dropped:', *old_unique, sep='\\n        ')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Show sample first line of data for each new file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output intro descriptions\n",
    "introText = (\n",
    "f'Sample first record from each datasource:\\n---\\n'\n",
    "f'    Somatic Alterations:        {NEW_OPENPEDCAN_SOMATIC_ALTERATIONS_VERSION:8}\\n'\n",
    "f'    Gene Expression:            {NEW_OPENPEDCAN_GENE_EXPRESSION_VERSION:8}\\n'\n",
    "f'    Epigenetic Modification:    {NEW_OPENPEDCAN_EPIGENETIC_MODIFICATION_VERSION:8}\\n')\n",
    "print(introText, '---\\n')\n",
    "\n",
    "# Iterate though data version groups\n",
    "for oldPath, newPath in pathTuples:\n",
    "\n",
    "    # Get filename(s) for output labeling\n",
    "    filename = newPath.rsplit('/',1)[-1]\n",
    "\n",
    "    print(f'{filename} ...\\n')\n",
    "\n",
    "    # If new file is missing, skip output\n",
    "    if not os.path.exists(newPath):\n",
    "        print(f'    File not found. \\n---\\n')\n",
    "        continue\n",
    "    \n",
    "    # Load first 10,000 lines of each dataset to gather columns\n",
    "    # (Loading more than 1 to avoid dropping columns with NaN values)\n",
    "    new = load_jsonl_as_chunks(newPath, limit=1, chunksize=1e4)\n",
    "\n",
    "    print(f'{new.iloc[0]} \\n---\\n')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
